epix/clustering.py

```diff
@@ -25,39 +26,75 @@ def find_cluster(interactions, cluster_size_space, cluster_size_time):
         df.append(ak.to_pandas(interactions[key], anonymous=key))
     df = pd.concat(df, axis=1)
 
-    # Splitting into individual events and cluster:
+    # Splitting into individual events and apply time clustering:
     groups = df.groupby('entry')
-    groups = groups.apply(lambda x: _find_cluster(x, cluster_size_space, cluster_size_time))
+    df["time_cluster"] = np.concatenate(groups.apply(lambda x: simple_1d_clustering(x.t.values, cluster_size_time)))
+
+    # Splitting into individual events and time cluster and apply space clustering space:
+    groups = df.groupby(['entry', 'time_cluster'])
+    groups = groups.apply(lambda x: _find_cluster(x, cluster_size_space))
+
+    for i in np.unique(groups.index.get_level_values(0)):
+        add_to_cluster = 0
+        for j in range(len(groups[i])):
+            groups[i][j]+=add_to_cluster
+            add_to_cluster = np.max(groups[i][j])+1
+
     df['cluster_id'] = np.concatenate(groups.values)
 
     ci = df.loc[:, 'cluster_id'].values
     offsets = ak.num(interactions['x'])
     interactions['cluster_ids'] = reshape_awkward(ci, offsets)
-
+    
     return interactions
 
+@numba.jit(nopython=True)
+def simple_1d_clustering(data, scale):
+    """
+    Function to cluster one dimensional data.
 
-def _find_cluster(x, cluster_size_space, cluster_size_time):
+    Args:
+        data (numpy.array): one dimensional array to be clusterd
+        scale (float): Max distance between two points to
+            be inside a cluster.
+    
+    Returns:
+        clusters_undo_sort (np.array): Cluster Labels
+    """
+    
+    idx_sort = np.argsort(data)
+    idx_undo_sort = np.argsort(idx_sort)
```

Comment:
I do not get this. Should not this be equivalent to `np.arange(0, len(data))`?

Reply:
Most likely this sort step is not needed since the event times are already sorted in time (I think). In case we give unsorted data to this function, we need to undo the sorting for the cluster labels to achieve a 1:1 correspondence between the produced cluster labels and the given input data.